internal/compiler-bridge/src/main/scala_2.13/xsbt/Compat.scala

```diff
@@ -28,6 +28,15 @@ abstract class Compat {
       func(a.original)
     }
   }
+
+  protected def processSAMAttachment(f: Function)(addInheritanceDependency: Symbol => Unit)
+      : Unit = {
+    f.attachments.get[SAMFunction].foreach { sam =>
+      {
+        addInheritanceDependency(sam.samTp.typeSymbol)
```

Comment:
I mainly want to request for feedback for properly registering inheritance relationship., As, the existing logic to register inheritance relationship has a rich history behind for which I don't fully understand.

```scala
      case Template(parents, self, body) =>
        // use typeSymbol to dealias type aliases -- we want to track the dependency on the real class in the alias's RHS
        def flattenTypeToSymbols(tp: Type): List[Symbol] =
          if (tp eq null) Nil
          else
            tp match {
              // rt.typeSymbol is redundant if we list out all parents, TODO: what about rt.decls?
              case rt: RefinedType => rt.parents.flatMap(flattenTypeToSymbols)
              case _               => List(tp.typeSymbol)
            }

        val inheritanceTypes = parents.map(_.tpe).toSet
        val inheritanceSymbols = inheritanceTypes.flatMap(flattenTypeToSymbols)

        debuglog(
          "Parent types for " + tree.symbol + " (self: " + self.tpt.tpe + "): " + inheritanceTypes + " with symbols " + inheritanceSymbols
            .map(_.fullName)
        )

        inheritanceSymbols.foreach { symbol =>
          addInheritanceDependency(symbol)
          addDependency(symbol)
        }

        inheritanceTypes.foreach(addTypeDependencies)
        addTypeDependencies(self.tpt.tpe)

        traverseTrees(body)
```

Reply:
@Friendseeker I have never looked into zinc's dependency tracking, so not speaking out of experience.

I found some documentation in [Relations.scala](https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala). Incremental compilation returns an `Analysis` containing the dependency analysis, the `relations: Relations` field is where the depdencies are represented.

It seems worth reading through that file. The top-level comment explains "internal" vs "external" vs "library" deps. The class has fields
  - `memberRef`: "dependency relation between classes introduced by member reference"
  - `inheritance`: "class dependency relation between classes introduced by inheritance"
  - `localInheritance`: see example in comment

See also [DependencyContext.java](https://github.com/sbt/zinc/blob/develop/internal/compiler-interface/src/main/contraband-java/xsbti/api/DependencyContext.java).

So it sounds like "inheritance dependencies" is a subset of "dependencies".

`addTypeDependencies` despite its similar name to `addDependency`/ `addInheritanceDependency`  is not a different kind of dependency registered in the analysis. It's a method that traverses all components of a `Type` and ultimately invokes `addDependency` for symols found within that type.

I guess this all means that we should also call `addDependency(sam.samTp.typeSymbol)` here. Again I really don't have experience, so I'm not sure. Maybe you can compare dependencies registered for `val f: () => Int = () => 1` vs `val f: A = () => 1`? [DependencySpecification.scala](https://github.com/sbt/zinc/blob/develop/internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/DependencySpecification.scala) might be a good place to implement that.

Reply:
Thanks for the detailed explanation! I will look into comparing dependencies registered in [DependencySpecification.scala](https://github.com/sbt/zinc/blob/develop/internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/DependencySpecification.scala).